File: 547/friend-circles.cpp
class Solution
{
  public:
    int findCircleNum(vector<vector<int>> &M)
    {
        int n = M.size(), res = 0;
        vector<bool> visited(n, false);
        for (int i = 0; i < n; i++)
        {
            if (visited[i])
                continue;
            helper(M, i, visited);
            res++;
        }
        return res;
    }
    void helper(vector<vector<int>> &M, int k, vector<bool> &visited)
    {
        visited[k] = true;
        for (int i = 0; i < M.size(); i++)
        {
            if (!M[k][i] || visited[i])
                continue;
            helper(M, i, visited);
        }
    }
};

File: 547/friend-circles.py
class Solution:
    def findCircleNum(self, M):
        """
        :type M: List[List[int]]
        :rtype: int
        """
        n = len(M)
        ans = 0
        hash = [False] * n

        def dfs(idx):
            hash[idx] = True
            for index, item in enumerate(M[idx]):
                if hash[index] == False and item == 1:
                    dfs(index)

        for i in range(n):
            if hash[i] == False:
                ans += 1
                dfs(i)

        return ans
